### PR TITLE
bundle: move tree-sitter as a bundled dep

### PR DESCRIPTION
**Pull request by vigoux**
Move tree-sitter as a bundled dependency.

EDIT: This works, but I don't know the robustness of it.

### CODE REVIEW

File: third-party/cmake/BuildTreesitter.cmake [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  -DDOWNLOAD_DIR=${DEPS_DOWNLOAD_DIR}/tree-sitter
+  -DURL=${TREESITTER_URL}
+  -DEXPECTED_SHA256=${TREESITTER_SHA256}
+  -DTARGET=tree-sitter
+  -DUSE_EXISTING_SRC_DIR=${USE_EXISTING_SRC_DIR}
+  -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
+  BUILD_IN_SOURCE 1
+  PATCH_COMMAND ""
+  CONFIGURE_COMMAND ""
+  INSTALL_COMMAND "PREFIX=${DEPS_INSTALL_DIR}" "make" "install")
```
Comment (vigoux) [inline third-party/cmake/BuildTreesitter.cmake:16]: I don't know any other way to do it, but I am sure there is one.

Comment (clason) [reply third-party/cmake/BuildTreesitter.cmake]: Just a heads-up: I don't think the empty commands are robust in cmake 3.18 -- can you try with that version?

Comment (vigoux) [reply third-party/cmake/BuildTreesitter.cmake]: Empty command just does nothing, which is not what I want I guess

Comment (clason) [reply third-party/cmake/BuildTreesitter.cmake]: That seems to have changed in 3.18, at least in some cases.

Comment (clason) [reply third-party/cmake/BuildTreesitter.cmake]: (see #12676 )

Comment (clason) [reply third-party/cmake/BuildTreesitter.cmake]: Never mind, I just tested with CMake 3.18 (macOS) and it works.

File: .travis.yml [hunk 2]
```diff
@@ -36,7 +36,7 @@ env:
                    -DBUSTED_OUTPUT_TYPE=nvim
                    -DDEPS_PREFIX=$DEPS_BUILD_DIR/usr
                    -DMIN_LOG_LEVEL=3"
-    - DEPS_CMAKE_FLAGS="-DUSE_BUNDLED_GPERF=OFF"
+    - DEPS_CMAKE_FLAGS="-DUSE_BUNDLED_GPERF=OFF -DUSE_BUNDLED_TS=ON"
```
Comment (jamessan) [inline .travis.yml:39]: This shouldn't be needed, since the default is to build all of third-party.

**Comment by vigoux** [comment]
Here is the problem, but I am not good enough to know how to fix this : https://travis-ci.org/github/neovim/neovim/jobs/728252354#L7793

**Comment by bfredl** [comment]
the problem is that CMakeLists want `/home/travis/nvim-deps/usr/lib/libtree-sitter.so` but it should want `/home/travis/nvim-deps/usr/lib/libtree-sitter.a` (bundled deps are static).

**Comment by jamessan** [comment]
> the problem is that CMakeLists want `/home/travis/nvim-deps/usr/lib/libtree-sitter.so` but it should want `/home/travis/nvim-deps/usr/lib/libtree-sitter.a` (bundled deps are static).

This happened because `BuildTreesitter.cmake` wasn't registering itself into our `THIRD_PARTY_DEPS` variable.  We use this to determine what shared libraries to remove from `.deps/usr/lib`, so that only the static library remains.

If CMake sees both a shared and static library, it will link against the shared library (since we haven't told it otherwise).

As for the Appveyor issue, it looks like Strawberry Perl's site is having problems.  Since we're mucking around in third-party/, the deps cache is invalidated and Appveyor isn't able to reliably download/install Strawberry Perl (which comes with cpanm).

**Comment by vigoux** [comment]
AppVeyor needs a fix upstream, I'll open a PR later next week.

**Comment by vigoux** [comment]
I _think_ that this is ready to merge, and that the AppVeyor error is just unrelated. The strange part is that only 1 test (asserting basic functionnality) fails, and every other one passes. I thus really think this is ready to merge.

**Comment by bfredl** [comment]
`Passed in:(nil)` is sometimes a secret code-word for a crash. Let's try once more..

**Comment by vigoux** [comment]
Windows failure is a complete mistery, as every other test is passing succesfully.

File: third-party/CMakeLists.txt [hunk 3]
```diff
@@ -198,6 +199,9 @@ set(LIBICONV_SHA256 ccf536620a45458d26ba83887a983b96827001e92a13847b45e4925cc891
 set(TREESITTER_C_URL https://github.com/tree-sitter/tree-sitter-c/archive/6002fcd.tar.gz)
 set(TREESITTER_C_SHA256 46f8d44fa886d9ddb92571bb6fa8b175992c8758eca749cb1217464e512b6e97)
 
+set(TREESITTER_URL https://github.com/tree-sitter/tree-sitter/archive/0.17.1.zip)
```
Comment (theHamsta) [inline third-party/CMakeLists.txt:202]: Have you checked whether the runtime update resolve the issue with newest Python/Go/Java parsers (they are locked to the revisions before the 0.17.1 CLI update)? Can parsers generated with old tree-sitter CLI (0.16) still be used without problems on the new runtime?

Comment (vigoux) [reply third-party/CMakeLists.txt:202]: I did not tried, but I guess it would be okay to use older generated parsers. This needs testing.

**Comment by vigoux** [comment]
I am completely unable to reproduce the windows issue localy, which is really weird.
I think I'll mark the test as flaky on windows for now.

**Comment by vigoux** [comment]
AppVeyor and freebsd fails are not related, IMO, can you confirm that @bfredl ?

**Comment by bfredl** [comment]
> AppVeyor and freebsd fails are not related, IMO, can you confirm that @bfredl ?

Yes.

**Comment by dominiklohmann** [comment]
Heads up: this broke `brew install --HEAD neovim`.

**Comment by bfredl** [comment]
@dominiklohmann actually posting, like logs or error messages, would be very useful.

**Comment by dominiklohmann** [comment]
@bfredl sorry, didn't mean to be rude. Here is the relevant excerpt—it can't find tree-sitter.

```
HEAD is now at a766695c8 Merge pull request #12931 from vigoux/treesitter-bundled
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20201103-2460-vpr3wh/deps-build
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20201103-2460-vpr3wh/deps-build
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20201103-2460-vpr3wh/deps-build
==> cmake . -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_INSTALL_PREFIX=/privat
==> make install
==> cmake .. -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DLIBLUV_INCLUDE_DIR=/private
==> make install
Last 15 lines from /Users/dominiklohmann/Library/Logs/Homebrew/neovim/07.make:
1 error generated.
[  0%] Generating auto/mbyte.c.generated.h, ../../include/mbyte.h.generated.h
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/bin/clang /tmp/neovim-20201103-2460-vpr3wh/src/nvim/mbyte.c -E -o /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/mbyte.i -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -DDO_NOT_DEFINE_EMPTY_ATTRIBUTES -I/tmp/neovim-20201103-2460-vpr3wh/build/config -I/tmp/neovim-20201103-2460-vpr3wh/src -I/usr/local/include -I/usr/local/include -I/tmp/neovim-20201103-2460-vpr3wh/deps-build/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk/usr/include -I/usr/local/opt/gettext/include -I/tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto -I/tmp/neovim-20201103-2460-vpr3wh/build/include -I/usr/local/include/luajit-2.0 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk -DNDEBUG -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
make[2]: *** [src/nvim/auto/lua/treesitter.c.generated.h] Error 1
[  0%] Generating auto/memfile.c.generated.h, ../../include/memfile.h.generated.h
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/bin/clang /tmp/neovim-20201103-2460-vpr3wh/src/nvim/memfile.c -E -o /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/memfile.i -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -DDO_NOT_DEFINE_EMPTY_ATTRIBUTES -I/tmp/neovim-20201103-2460-vpr3wh/build/config -I/tmp/neovim-20201103-2460-vpr3wh/src -I/usr/local/include -I/usr/local/include -I/tmp/neovim-20201103-2460-vpr3wh/deps-build/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk/usr/include -I/usr/local/opt/gettext/include -I/tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto -I/tmp/neovim-20201103-2460-vpr3wh/build/include -I/usr/local/include/luajit-2.0 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk -DNDEBUG -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/local/bin/luajit /tmp/neovim-20201103-2460-vpr3wh/src/nvim/generators/gen_declarations.lua /tmp/neovim-20201103-2460-vpr3wh/src/nvim/math.c /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/math.c.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/include/math.h.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/math.i
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/local/bin/luajit /tmp/neovim-20201103-2460-vpr3wh/src/nvim/generators/gen_declarations.lua /tmp/neovim-20201103-2460-vpr3wh/src/nvim/main.c /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/main.c.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/include/main.h.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/main.i
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/local/bin/luajit /tmp/neovim-20201103-2460-vpr3wh/src/nvim/generators/gen_declarations.lua /tmp/neovim-20201103-2460-vpr3wh/src/nvim/map.c /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/map.c.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/include/map.h.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/map.i
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/local/bin/luajit /tmp/neovim-20201103-2460-vpr3wh/src/nvim/generators/gen_declarations.lua /tmp/neovim-20201103-2460-vpr3wh/src/nvim/marktree.c /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/marktree.c.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/include/marktree.h.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/marktree.i
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/local/bin/luajit /tmp/neovim-20201103-2460-vpr3wh/src/nvim/generators/gen_declarations.lua /tmp/neovim-20201103-2460-vpr3wh/src/nvim/mark.c /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/mark.c.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/include/mark.h.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/mark.i
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/local/bin/luajit /tmp/neovim-20201103-2460-vpr3wh/src/nvim/generators/gen_declarations.lua /tmp/neovim-20201103-2460-vpr3wh/src/nvim/memfile.c /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/memfile.c.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/include/memfile.h.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/memfile.i
cd /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim && /usr/local/bin/luajit /tmp/neovim-20201103-2460-vpr3wh/src/nvim/generators/gen_declarations.lua /tmp/neovim-20201103-2460-vpr3wh/src/nvim/mbyte.c /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/mbyte.c.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/include/mbyte.h.generated.h /tmp/neovim-20201103-2460-vpr3wh/build/src/nvim/auto/mbyte.i
make[1]: *** [src/nvim/CMakeFiles/nvim.dir/all] Error 2
make: *** [all] Error 2

Do not report this issue to Homebrew/brew or Homebrew/core!
```

**Comment by theHamsta** [comment]
There are your testers :smile:! As we sad, merging will bring them.

**Comment by clason** [comment]
So the brew build script needs to be adapted. (Building from master works fine on macOS with the included scripts, so `make`.)

The issue is that homebrew does not use bundled dependencies, rather relying on their own versions where possible. This means that they now need to add a tree-sitter dependency, or edit their build script to use the bundled version here.

(And by "they should edit", I mean "somebody has to make a PR" -- note that this only affects `HEAD`, which isn't fully supported, so don't expect a quick fix.)

**Comment by vigoux** [comment]
I have a brew on a machine, so I'll try my best to make the PR, but don't expect anything.

**Comment by dominiklohmann** [comment]
I can take a stab at this myself after work today as well if you need help. tree-sitter is not currently a formula in homebrew-core, so it'd need to be added to that repository. The neovim formula must depend on tree-sitter when building the formula from HEAD.

If someone's reading this: As a workaround, build and install tree-sitter from source before running `brew install --HEAD neovim`.

**Comment by vigoux** [comment]
I'll let you do it then if you want, I am not familiar at all with homebrew formulas.
Feel free to ping me if you need any help

**Comment by clason** [comment]
> If someone's reading this: As a workaround, build and install tree-sitter from source before running `brew install --HEAD neovim`.

In this case, the easiest thing is probably to add a tree-sitter formula and depend on that.

**Comment by clason** [comment]
You don't even have to start from scratch: https://github.com/ChimeHQ/homebrew-tree-sitter

**Comment by vigoux** [comment]
I think there already is `tree-sitter-cli`, so this should become `libtree-sitter` (which is the actual name of the library), and maybe provide `tree-sitter` as both of them.

**Comment by dominiklohmann** [comment]
Opened a PR at https://github.com/Homebrew/homebrew-core/pull/64015

**Pull request merged**